### DISCUSSION TITLE
Agent-level concurrency limitation for docker-based agents in Cloud Hybrid

### DISCUSSION THREAD

**Discussion by mudravrik** [dagster-io/dagster#21214] [Q&A] [answered]
Hi all!

I wonder if there is a way to limit how many runs can be sent to one particular Docker agent connected to our Dagster Hybrid Cloud.

Our usecase is pretty simple: 95% of the time we have very low workload, so it can be managed with pretty small agent machine.
However we occasionally perform huge loads, which requires way more processing power.

So my initial idea was to bring additional agent dedicated for this workloads and take it down as soon as load finished.
However I'm not quite sure how to manage jobs between agents to prevent accidental OOM-ing/killing smaller "day-to-day" agent.
I assumed there should be some way to limit number of jobs per agent, or make agent a priority target for some of the jobs by tags, but I cannot find anything related in docs or source code.
So hoping for any help here :)

Ps. I understand, that using K8s-based agents is, probably, a proper way to handle such computation scaling, however dealing with increased infrastructure complexity is not worth it right now for us.

**Comment by mlarose** [comment, marked as answer]
Hi @mudravrik,

there's no agent scoped run limit unfortunately. 

> So my initial idea was to bring additional agent dedicated for this workloads and take it down as soon as load finished.

Unless you're taking down your small agent while your big agent is up, you would have to use agent queues to control which agents gets your job, which basically map a code location to a queue and by configuration agent can decide which queue they consume. Now that implies you can separate your huge workload from your small ones.

Then, if you can do that you might be able to achieve this by [limiting concurrency](https://docs.dagster.io/guides/limiting-concurrency-in-data-pipelines) at an appropriate level depending on your need.

Or a simpler path with some manual steps each time. Adjust [concurrency globally](https://docs.dagster.io/guides/limiting-concurrency-in-data-pipelines#limiting-overall-runs). 

IE, spin down small agent, spin down big agent, bump overall concurrency to whatever is needed, execute bulk jobs, then revert/scale down.

Both workarounds, seems either tedious or just another way to build up the complexity of your platform, so I am not particularly pleased to offer these avenues.


> Ps. I understand, that using K8s-based agents is, probably, a proper way to handle such computation scaling, however dealing with increased infrastructure complexity is not worth it right now for us.

You could also consider the ECS agent and achieve the same scalability goal for less time investment. k8s is more powerful but ECS+Fargate can definitely give you some decent mileage for relatively low effort.

**Reply by mudravrik** [reply]
@mlarose Thank you for the reply!

> you would have to use agent queues to control which agents gets your job, which basically map a code location to a queue and by configuration agent can decide which queue they consume.

It actually will work fine for us. We already have all heavy-lifting jobs separated in one Code Location, so if we could send all runs from this location to the dedicated agent, it will be great.
However, I cannot find anything related to "agent queues" in documentation you linked. Maybe you can point me to some detailed guide on it?

> You could also consider the ECS agent and achieve the same scalability goal for less time investment. k8s is more powerful but ECS+Fargate can definitely give you some decent mileage for relatively low effort.

Unfortunately, we live on GCP and I'm pretty sure attempt of use something like Cloud Run without detailed docs like for Docker/ECG/K8s will be way harder to implement for us :)

**Comment by mlarose** [comment]
> However, I cannot find anything related to "agent queues" in documentation you linked. Maybe you can point me to some detailed guide on it?

Indeed, I forgot to link that other concept.

This section of the documentation is what you are looking for:

https://docs.dagster.io/dagster-cloud/deployment/agents/running-multiple-agents#routing-requests-to-specific-agents

> Unfortunately, we live on GCP and I'm pretty sure attempt of use something like Cloud Run without detailed docs like for Docker/ECG/K8s will be way harder to implement for us :)

yes indeed, creating a custom agent is heavy lifting.

**Reply by mudravrik** [reply]
Thank you so much, I'm gonna try it!
